freqtrade/freqtradebot.py

```diff
@@ -479,12 +462,33 @@ def process_maybe_execute_buys(self) -> None:
         """
         Tries to execute buy orders for trades in a safe way
         """
-        try:
-            # Create entity and execute trade
-            if not self.create_trades():
-                logger.debug('Found no buy signals for whitelisted currencies. Trying again...')
-        except DependencyException as exception:
-            logger.warning('Unable to create trade: %s', exception)
+        result = False
+
+        whitelist = copy.deepcopy(self.active_pair_whitelist)
+        if not whitelist:
+            logger.info("Active pair whitelist is empty.")
+        else:
+            # Remove currently opened and latest pairs from whitelist
+            for trade in Trade.get_open_trades():
+                if trade.pair in whitelist:
+                    whitelist.remove(trade.pair)
+                    logger.debug('Ignoring %s in pair whitelist', trade.pair)
+
+            if not whitelist:
+                logger.info("No currency pair in active pair whitelist, "
+                            "but checking to sell open trades.")
+            else:
+                # Create entity and execute trade for each pair from whitelist
+                for pair in whitelist:
+                    try:
+                        if self.create_trade(pair):
+                            result = True
+                    except DependencyException as exception:
+                        logger.warning('Unable to create trade: %s', exception)
```

Comment:
I'm not certain about handling DependencyExceptions "per pair", and i think there's 2 sides to this, depending on where the exception "came from".

One side is what you mention in the Pr comment - that one failed order should not cause the buy loop to stop.

However, if the exception came from `_check_available_stake_amount()` (for example) - then it will raise again for every pair the bot tries to buy - because stake-amount won't "magically" become available within the iteration - so i would suddenly encounter 20 messages instead of one (before this PR) - each telling me that it could not buy.

Yes, you could add balance to the exchange so it's a theoretical possibility - but i don't think that needs to be handled within one iteration.

We'll probably need more fain-grained Exceptions, where we can control "this interrupts the buy-loop" - and "this other one only for this pair" - but that's clearly out of scope for this PR.

Reply:
> We'll probably need more fain-grained Exceptions, where we can control "this interrupts the buy-loop" - and "this other one only for this pair" - but that's clearly out of scope for this PR.

Yes, this is how exceptions should work here and should be handled here: DependencyExceptionPair (whatever suitable name can be used) is catched inside the loop and terminates handling of the current trade (pair) here, DependencyExceptionIteration catched outside of the loop and terminates the whole 'buy' part of the business logic, OperationalException terminates the bot completely since signals non-resurrectable conditions...
Not in this PR.

Reply:
definitely not - i started with it in #2726 - but intentionally did not add any new exceptions at a "raising" level ... since i guess we'll need a few more exceptions, and we probably should add them when we need them ...